### PR TITLE
T2D-169 Get Accession History

### PR DESCRIPTION
**Pull request by jorizci**
Based on #35 
- Rest package reorder
- Model / Service / repository package reorder
- Removed unused classes due to changes
- Added History service

### CODE REVIEW

File: accession-commons-mongodb/src/test/java/uk/ac/ebi/ampt2d/test/HistoryTester.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                .andThen(assertVersion(version)).andThen(assertEventContentIs(value));
+    }
+
+    private static Consumer<HistoryEvent<TestModel, String>> assertEventContentIs(String value) {
+        return event -> assertEquals(value, event.getData().getValue());
+    }
+
+    public static Consumer<HistoryEvent<TestModel, String>> assertEventIsPatch(String value, int version) {
+        return assertEventIs(EventType.PATCHED).andThen(assertNullMergedInto()).andThen(assertVersion(version))
+                .andThen(assertVersion(version)).andThen(assertEventContentIs(value));
```
Comment (selvaebi) [inline accession-commons-mongodb/src/test/java/uk/ac/ebi/ampt2d/test/HistoryTester.java]: assertVersion(version) is checked twice here.

File: accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/core/models/HistoryEvent.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    private Integer version;
+
+    private ACCESSION mergedInto;
+
+    private LocalDateTime localDateTime;
+
+    private MODEL data;
+
+    public HistoryEvent(EventType type, ACCESSION accession, Integer version, ACCESSION mergedInto,
```
[the remark this answers is not in the document]
Comment (jorizci) [reply accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/core/models/HistoryEvent.java]: Only in the merge and I don't think the database uses that nomenclature for accession.

File: accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/models/IOperation.java [hunk 3]
```diff
@@ -15,26 +15,31 @@
...
+import java.util.List;
 
 /**
  * An entry of operations performed over accessioned objects that are considered no longer active.
+ *
+ * @param <MODEL>
  * @param <ACCESSION>
  */
-public interface IOperation<ACCESSION> {
+public interface IOperation<MODEL, ACCESSION> {
```
[the remark this answers is not in the document]
Comment (jorizci) [reply accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/models/IOperation.java]: Done

File: accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/services/BasicHistoryService.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            switch (operation.getEventType()) {
+                case DEPRECATED:
+                    versionMap = mapVersions(operation.getInactiveObjects());
+                    newEvent = HistoryEvent.deprecated(operation.getAccession(), operation.getCreatedDate());
+                    lastVersion = versionMap.keySet().size();
+                    break;
+                case MERGED:
+                    versionMap = mapVersions(operation.getInactiveObjects());
+                    newEvent = HistoryEvent.merged(operation.getAccession(),
+                            operation.getMergedInto(), operation.getCreatedDate());
```
Comment (selvaebi) [inline accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/services/BasicHistoryService.java]: I think we also need to initialize lastVersion same like deprecated here;

File: accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/dto/HistoryEventDTO.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    private LocalDateTime localDateTime;
+
+    private DTO data;
+
+    public <MODEL> HistoryEventDTO(HistoryEvent<MODEL, ACCESSION> event, Function<MODEL, DTO> modelToDTO) {
+        this.type = event.getType();
+        this.accession = event.getAccession();
+        this.mergedInto = event.getMergedInto();
+        this.localDateTime = event.getLocalDateTime();
+        this.data = modelToDTO.apply(event.getData());
```
Comment (cyenyxe) [inline accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/dto/HistoryEventDTO.java]: The conversion from entity to model now works differently than the conversion from model to DTO. This one uses a `Function` whereas the other is a direct cast.

File: accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/dto/HistoryEventDTO.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ */
+package uk.ac.ebi.ampt2d.commons.accession.rest.dto;
+
+import uk.ac.ebi.ampt2d.commons.accession.core.models.EventType;
+import uk.ac.ebi.ampt2d.commons.accession.core.models.HistoryEvent;
+
+import java.time.LocalDateTime;
+import java.util.function.Function;
+
+public class HistoryEventDTO<ACCESSION, DTO> {
```
[the remark this answers is not in the document]
Comment (jorizci) [reply accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/dto/HistoryEventDTO.java]: Not really needed for anything as this is only to read.

File: README.md [hunk 7]
```diff
@@ -1,5 +1,5 @@
 # Accession Commons
-Generic library for accessioning any object type. It provides the core interfaces to generate hash-based and monotonically increasing accessions, which must be extended by each application to support a particular object type.
+Generic library for accessioning any object eventType. It provides the core interfaces to generate hash-based and monotonically increasing accessions, which must be extended by each application to support a particular object eventType.
```
[the remark this answers is not in the document]
Comment (jorizci) [reply README.md]: IntelliJ 🙄

File: accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/core/models/IEvent.java [hunk 8]
```diff
@@ -15,26 +15,31 @@
...
 
-import uk.ac.ebi.ampt2d.commons.accession.core.OperationType;
+import uk.ac.ebi.ampt2d.commons.accession.persistence.models.IAccessionedObject;
 
 import java.time.LocalDateTime;
+import java.util.List;
 
 /**
- * An entry of operations performed over accessioned objects that are considered no longer active.
+ * An entry of operations performed over accessioned objects
```
[the remark this answers is not in the document]
Comment (jorizci) [reply accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/core/models/IEvent.java]: I don't see it much more clear :/

File: accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/dto/HistoryEventDTO.java (hunk 5 above)
Comment (jorizci) [reply accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/dto/HistoryEventDTO.java]: We can't do a direct cast for the DTO. We need to use the DTO object if we want to use them to define the mapping for the json objects.

**Comment by jorizci** [comment]
Is the current conversion model doesn't seem as easy as the previous one I can roll back that change.

**Comment by cyenyxe** [comment]
I don't mean the current change, but I feel the model as a whole is becoming quite complicated with the combinations of entities, models, DTOs, wrappers and different ways to convert between them.

**Pull request merged**